source/vibe/web/rest.d

```diff
@@ -618,6 +660,88 @@ private HTTPServerRequestDelegate jsonMethodHandler(alias Func, size_t ridx, T)(
 	return &handler;
 }
 
+/**
+ * Generate an handler that will wrap the server's method
+ *
+ * This function returns an handler that handles the http OPTIONS method.
+ *
+ * It will return the ALLOW header with all the methods on this resource
+ * And it will handle Preflight CORS.
+ *
+ * Params:
+ *	routes = a range of Routes were each route has the same resource/URI
+ *				just different method.
+ *	settings = REST server configuration.
+ *
+ * Returns:
+ *	A delegate suitable to use as an handler for an HTTP request.
+ */
+private HTTPServerRequestDelegate optionsMethodHandler(RouteRange)(RouteRange routes, RestInterfaceSettings settings = null)
+{
+	import vibe.http.server : HTTPServerRequest, HTTPServerResponse;
+	import std.algorithm : map, joiner, any;
+	import std.conv : text;
+	import std.array : array;
+	import vibe.http.common : httpMethodString, httpMethodFromString;
+	// NOTE: don't know what is better, to keep this in memory, or generate on each request
+	auto allow = routes.map!(r => r.method.httpMethodString).joiner(",").text();
+	auto methods = routes.map!(r => r.method).array();
+
+	void handlePreflightedCors(HTTPServerRequest req, HTTPServerResponse res, ref HTTPMethod[] methods, RestInterfaceSettings settings = null)
+	{
+		import std.algorithm : among;
+
+		auto origin = "Origin" in req.headers;
+		if (origin is null)
+			return;
+
+		import std.algorithm : equal;
+		import std.uni : toLower;
+		auto originLowerCased = (*origin).toLower;
+		if (settings !is null &&
+			settings.allowedOrigins.length != 0 &&
+			!settings.allowedOrigins.any!(org => org.toLower.equal(originLowerCased)))
+			return;
+		
+		auto method = "Access-Control-Request-Method" in req.headers;
+		if (method is null)
+			return;
+
+		auto httpMethod = httpMethodFromString(*method);
+
+		if (!methods.any!(m => m == httpMethod))
+			return;
+
+		res.headers["Access-Control-Allow-Origin"] = *origin;
+
+		// there is no way to know if the specific resource supports credentials
+		// (either cookies, HTTP authentication, or client-side SSL certificates), 
+		// so we always assume it does
+		res.headers["Access-Control-Allow-Credentials"] = "true";
+		res.headers["Access-Control-Max-Age"] = "1728000";
+		res.headers["Access-Control-Allow-Methods"] = *method;
```

Comment:
Its funny, elsewhere I was complaining that you don't get access to the raw request/response headers. Now it is obvious you don't want that. In fact, any dealings with low level http details should be know at compile time (at least with the rest interface)

Having said that, its going to take a long time before anybody complains about allow-credentials being non-configurable. Maybe max-age...